managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java

```diff
@@ -80,24 +105,37 @@ public int getLength() {
 
     @Override
     public PositionImpl getPosition() {
-        return position;
+        // return new non-recyclable instance of position as this.position will be recycled once (EntryImpl)this will be
+        // recycled. So, return new instance whose life-cycle doesn't depend on EntryImpl
+        return new PositionImpl(position);
     }
 
     @Override
     public int compareTo(EntryImpl other) {
         return position.compareTo(other.getPosition());
     }
 
+    @Override
     public void retain() {
         data.retain();
     }
 
     @Override
     public void release() {
-        data.release();
+        if(data.release()) {
+            // recycle only if data buf is released and no other object will use the data 
+            recycle();    
+        }
     }
 
-    public int refCnt() {
-        return data.refCnt();
+    @Override
+    public void recycle() {
+        this.data = null;
+        // recycle position if it is recyclable object
+        this.position.recycle();
```

Comment:
>Recyclying on the PositionImpl is currently disabled (and actually it might be good to remove that dead-code in this PR as well)

I have already cleaned up dead-code and enabled the recycling on [PositionImpl](https://github.com/rdhabalia/pulsar/blob/f3199810d6997e026a4dd0dc0bbf6593e0d327a2/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionImpl.java)

>we can reuse the same position all the time, we just need to reset the ledgerId and entryId on it.

Don't you think recycle would be better as it can protect to mess up things and it will surely return null on already recycled entryImpl.getPosition() rather reset value.